Custom Ingress Path for Helm Chart

JordanZimmitti opened the pull request:
### Description
Now that Selenium supports custom sub-paths for the UI, in order to take advantage of it in Kubernetes the helm chart needs the ingress `path` to be overwritable. I added a `path` key under the ingress section of the `values.yaml` file and made it backward compatible in the template so if no `path` key exists then it will use the default "/" path just like before.

### Motivation and Context
Deploy selenium with a custom domain sub-path using the helm chart

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the [contributing](https://selenium.dev/documentation/en/contributing/) document.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
<!--- Provide a general summary of your changes in the Title above -->

JordanZimmitti committed:
Added ingress path option to helm values file

JordanZimmitti committed:
Update README.md

JordanZimmitti committed:
Update README.md

JordanZimmitti committed:
Update README.md

JordanZimmitti committed:
Update ingress.yaml

JordanZimmitti committed:
Updated chart version

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/SeleniumHQ/docker-selenium?pullRequest=1834) <br/>All committers have signed the CLA.

diemol commented on charts/selenium-grid/Chart.yaml:
```diff
@@ -2,6 +2,6 @@ apiVersion: v2
 description: A Helm chart for creating a Selenium Grid Server in Kubernetes
 type: application
-version: 0.15.8
+version: 0.15.9
```
I would move this to 0.16.0 since there is a new feature.

JordanZimmitti replied:
Makes sense, I updated the value to 0.16.0

diemol reviewed (commented):
I do not see how the sub path is getting configured in the Grid itself. Am I missing something?

JordanZimmitti committed:
Update Chart.yaml

JordanZimmitti commented:
@diemol 
> I do not see how the sub path is getting configured in the Grid itself. Am I missing something?

I was able to configure that part using the existing helm chart already by adding these lines here:

```yaml
hub:
    extraEnvironmentVariables:
        - name: SE_OPTS
           value: '--sub-path /<custom-sub-path>/'
  
```

which worked fine, but then could not update the ingress path itself since it was hard coded

diemol commented:
@JordanZimmitti would if be better if we add a new environment variable just to set the sub-path? If so, I can have a look at it unless you want to do it.

JordanZimmitti commented:
@diemol it did take me a little while to figure out how to set the subpath for the hub so having an env called `SUB_PATH` or something similar might be easier for people to interpret. I'll take a stab at it I think I might know where that needs to be updated

JordanZimmitti committed:
Add explicit SE_HUB_SUB_PATH environment variable for the HUB

JordanZimmitti commented:
@diemol I updated the selenium hub start script to take a new environment variable called `SE_HUB_SUB_PATH` which will set the sub-path. Tested this locally with and without using the env variable and it worked as expected. I followed the naming convention that I saw being used in the script, if you would like something different let me know and I'll update it

diemol commented on Hub/start-selenium-grid-hub.sh:
```diff
@@ -17,6 +17,11 @@ if [ ! -z "$SE_HUB_PORT" ]; then
   PORT_CONFIG="--port ${SE_HUB_PORT}"
 fi
 
+if [ ! -z "$SE_HUB_SUB_PATH" ]; then
```
I would just use `SE_SUB_PATH`. Why?

Because we should also use it in the [Standalone](https://github.com/SeleniumHQ/docker-selenium/blob/trunk/Standalone/start-selenium-standalone.sh) and the [Router](https://github.com/SeleniumHQ/docker-selenium/blob/trunk/Router/start-selenium-grid-router.sh).

JordanZimmitti replied:
Sounds good will do!

diemol reviewed (changes requested):
Looks good, we just need to add it to a couple of places more. I'd be nice if you have time to put a couple of lines describing it on the main README. Thanks!

JordanZimmitti committed:
Add SE_SUB_PATH to recommended scripts and added documentation

JordanZimmitti commented:
@diemol Thanks for the suggestions! I updated the environment variable name to `$SE_SUB_PATH` as suggested and added it to the `Standalone` and `Router` scripts. I also added some documentation for using the environment variable in the `readme.md` in a spot that I think makes sense but if you want me to put it somewhere else let me know

diemol reviewed (commented):
Looks great!

Not sure if I am missing something, but shouldn't the `SE_SUB_PATH` variable be used somewhere in the chart?

JordanZimmitti commented:
@diemol 
> Looks great!
> 
> Not sure if I am missing something, but shouldn't the `SE_SUB_PATH` variable be used somewhere in the chart?

Since the docker file takes `SE_SUB_PATH` in as an environment variable it would work fine with the helm chart as is like this

```yaml
hub:
    extraEnvironmentVariables:
        - name: SE_SUB_PATH
           value: "/<custom-sub-path>/"
```

But I could break it out of `extraEnvironmentVariables` and make it its own key under the `hub` section if you want.
If we go down that path then I probably need to make a `sub_path` key under the `router` section too since we added it to the start script there as well correct?

diemol commented:
@JordanZimmitti sorry for the late reply, we had a long weekend here. 

>But I could break it out of extraEnvironmentVariables and make it its own key under the hub section if you want.
If we go down that path then I probably need to make a sub_path key under the router section too since we added it to the start script there as well correct?

I think ☝️ is the preferred solution so all places where the env var can be used get the option to be configured. Thanks!

JordanZimmitti commented:
@diemol Sounds good ill implement the changes in the next day or so!

JordanZimmitti committed:
Merge branch 'SeleniumHQ:trunk' into trunk

JordanZimmitti committed:
Added sub_path helm value for hub and components

JordanZimmitti commented:
@diemol 
I broke out the environment variable in the helm charts in the needed locations, let me know if you have any other suggestions!

diemol commented on charts/selenium-grid/templates/distributor-deployment.yaml:
```diff
@@ -38,6 +38,8 @@ spec:
             - name: SE_SESSION_QUEUE_PORT
               value: {{ .Values.components.sessionQueue.port | quote }}
+            - name: SE_SUB_PATH
+              value: {{ .Values.components.subPath }}
```
This is not needed here. The sub-path only applies to Router and Hub.

diemol commented on charts/selenium-grid/templates/event-bus-deployment.yaml:
```diff
@@ -36,9 +36,12 @@ spec:
-        {{- with .Values.components.extraEnvironmentVariables }}
-          env: {{- tpl (toYaml .) $ | nindent 12 }}
-        {{- end }}
+          env:
+            - name: SE_SUB_PATH
+              value: {{ .Values.components.subPath }}
+          {{- with .Values.components.extraEnvironmentVariables }}
+            {{- tpl (toYaml .) $ | nindent 12 }}
+          {{- end }}
```
This is not needed here. The sub-path only applies to Router and Hub.

diemol commented on charts/selenium-grid/templates/session-map-deployment.yaml:
```diff
@@ -29,9 +29,12 @@ spec:
-        {{- with .Values.components.extraEnvironmentVariables }}
-          env: {{- tpl (toYaml .) $ | nindent 12 }}
-        {{- end }}
+          env:
+            - name: SE_SUB_PATH
+              value: {{ .Values.components.subPath }}
+          {{- with .Values.components.extraEnvironmentVariables }}
+            {{- tpl (toYaml .) $ | nindent 12 }}
+          {{- end }}
```
This is not needed here. The sub-path only applies to Router and Hub.

diemol commented on charts/selenium-grid/templates/session-queuer-deployment.yaml:
```diff
@@ -29,9 +29,12 @@ spec:
-        {{- with .Values.components.extraEnvironmentVariables }}
-          env: {{- tpl (toYaml .) $ | nindent 12 }}
-        {{- end }}
+          env:
+            - name: SE_SUB_PATH
+              value: {{ .Values.components.subPath }}
+          {{- with .Values.components.extraEnvironmentVariables }}
+            {{- tpl (toYaml .) $ | nindent 12 }}
+          {{- end }}
```
This is not needed here. The sub-path only applies to Router and Hub.

diemol reviewed (changes requested):
Thank you!

I just noted that the sub-path environment variable is being used in a few places we do not need it.

JordanZimmitti committed:
Removed sub_path from components that do not use it

JordanZimmitti commented:
@diemol I removed the sub_path from components that don't need it based on your recommendation!

diemol committed:
Merge branch 'trunk' into trunk

diemol reviewed (approved):
Thank you, @JordanZimmitti!
